### ISSUE TITLE
assertion failed: ty.is_aggregate(context) when compiling

### ISSUE THREAD

**Issue by dmihal** [FuelLabs/sway#5377]
When I compile my contracts, I get the error:

```
thread 'main' panicked at sway-ir/src/optimize/sroa.rs:174:25:
assertion failed: ty.is_aggregate(context)
```

Message me for code to reproduce.

**Comment by ironcev** [comment]
@dmihal Please add the code to reproduce.

**Comment by ironcev** [comment]
While writing tests for new language functionality, references, I've stumbled upon the same issue. It looks like it has to have something with `u256` embedded in arrays, structs, or tuples.

I'll find some time to create a minimum repro. In the meantime, the TODOs in this file are showing the same issue: https://github.com/FuelLabs/sway/pull/5406/files#diff-840573b527b264b0015d3d6263fce47b06a9f36241172c730b77ee75af23f974

The parts of tests that are commented out are those failing with the same assertion.

CC: @vaivaswatha

**Comment by vaivaswatha** [comment]
I've assigned myself and waiting on @ironcev for a standalone repro.

**Comment by ironcev** [comment]
Thanks for taking care of the issue @vaivaswatha! Here are the min repros for arrays, structs, and tuples:

```Sway
script;

fn main() {
    let mut a = [0x0111111111111111111111111111111111111111111111111111111111111111u256];

    assert(a[0] == 0x0111111111111111111111111111111111111111111111111111111111111111u256);

    a[0] = 0x0222222222222222222222222222222222222222222222222222222222222222u256;
}
```
```Sway
script;

struct S {
    x: u256,
}

fn main() {
    let mut s = S { x: 0x0111111111111111111111111111111111111111111111111111111111111111u256 };

    assert(s.x == 0x0111111111111111111111111111111111111111111111111111111111111111u256);

    s.x = 0x0222222222222222222222222222222222222222222222222222222222222222u256;
}
```
```Sway
script;

fn main() {
    let mut t = (0x0111111111111111111111111111111111111111111111111111111111111111u256,);

    assert(t.0 == 0x0111111111111111111111111111111111111111111111111111111111111111u256);

    t.0 = 0x0222222222222222222222222222222222222222222222222222222222222222u256;
}
```

**Comment by ironcev** [comment]
I've just uncommented the [TODOs in the tests for dereferencing references](https://github.com/FuelLabs/sway/pull/5406/files#diff-840573b527b264b0015d3d6263fce47b06a9f36241172c730b77ee75af23f974) and the tests are passing fine. So, confirmed from my side :smile: Thanks for the prompt fix Vaivas!

**Comment by ironcev** [comment]
Uuups, just realized that merging the PR which I was eager to get so I can remove my TODOs also closed this issue without the confirmation from @dmihal that his code also works. David, out of our discussion I am sure that the fix will also solve your issue. In the unlikely case that it doesn't we can always reopen this issue so merging of the PR shouldn't be the problem on the contrary. I apologize for rushing...

**Comment by dmihal** [comment]
Tried it out, and it seems to solve the issue! 👍

**Comment by dmihal** [comment]
I'm running into this issue again on 0.46.1, I can share code if helpful

**Comment by dmihal** [comment]
Resolved in 0.56.0
